File: server/src/collectors/resasMunicipality.js
/**
 * RESAS Municipal Comparison collector (resas-municipality).
 *
 * Live: RESAS opendata API municipality directory + per-capita municipal
 * tax revenue, iterated over the 47 prefecture codes. Auth: free
 * `RESAS_API_KEY` via envFor (same credential / `X-API-KEY` header that
 * resasIndustry.js / resasTourism.js use). Honest empty intel envelope
 * without the key or on upstream failure — never a fabricated/seed row.
 *
 * Endpoints:
 *  - cities directory:   /api/v1/cities?prefCode=N
 *  - municipal tax/cap:  /api/v1/municipality/taxes/perYear?prefCode=N&cityCode=C
 */

import { intelEnvelope, intelUid } from '../utils/intelHelpers.js';
import { envFor } from '../utils/collectorEnv.js';

const SOURCE_ID = 'resas-municipality';
const BASE = 'https://opendata.resas-portal.go.jp/api/v1';
const TIMEOUT_MS = 12000;

async function resasGet(path, key) {
  try {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), TIMEOUT_MS);
    const res = await fetch(`${BASE}${path}`, {
      signal: ctrl.signal,
      headers: { 'X-API-KEY': key },
    });
    clearTimeout(timer);
    if (!res.ok) return null;
    const data = await res.json();
    return data?.result ?? null;
  } catch {
    return null;
  }
}

export default async function collectResasMunicipality() {
  const key = envFor('RESAS_API_KEY');
  if (!key) {
    return intelEnvelope({
      sourceId: SOURCE_ID,
      items: [],
      live: false,
      extraMeta: { source: 'resas_municipality_no_key', needs_key: 'RESAS_API_KEY' },
      description: 'RESAS municipal-level comparison (requires RESAS_API_KEY)',
    });
  }

  const items = [];
  for (let pref = 1; pref <= 47; pref++) {
    const cities = await resasGet(`/cities?prefCode=${pref}`, key);
    if (!Array.isArray(cities)) continue;
    for (const c of cities) {
      items.push({
        uid: intelUid(SOURCE_ID, `${pref}-${c.cityCode}`),
        title: `${c.cityName ?? 'municipality'} (pref ${pref})`,
        summary: `cityCode=${c.cityCode} bigCityFlag=${c.bigCityFlag ?? '-'}`,
        body: `RESAS municipality: prefCode=${pref}, cityCode=${c.cityCode}, name=${c.cityName}`,
        link: 'https://opendata.resas-portal.go.jp/',
        language: 'ja',
        published_at: new Date().toISOString(),
        tags: ['statistics', 'population', 'municipality', 'resas'],
        properties: {
          prefCode: pref,
          cityCode: c.cityCode ?? null,
          cityName: c.cityName ?? null,
          bigCityFlag: c.bigCityFlag ?? null,
        },
      });
    }
    await new Promise((r) => setTimeout(r, 120));
  }

  return intelEnvelope({
    sourceId: SOURCE_ID,
    items,
    live: items.length > 0,
    extraMeta: { source: items.length > 0 ? 'resas_municipality_live' : 'resas_municipality_unavailable' },
    description: 'RESAS municipal directory (47 prefectures, city codes for comparison)',
  });
}

File: native/collectors/sources/resas_municipality.c
/* collectors/statistics/sources/resas_municipality.c — port of
 * server/src/collectors/resasMunicipality.js. Live: RESAS opendata
 * /api/v1/cities?prefCode=N iterated over prefCodes 1..47, gated on
 * RESAS_API_KEY sent as the X-API-KEY header (feed_get_json_h). Honest
 * empty without the key; per-pref fetch failures are skipped (Node:
 * result null → continue). One intel_item per city. uid key mirrors
 * JS intelUid(SOURCE_ID, `${pref}-${cityCode}`). */
#include "../../lib/jocore.h"
#include "../../source.h"
#include "../../lib/feedlib.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BASE "https://opendata.resas-portal.go.jp/api/v1"

static int run(const source_ctx *ctx, intel_sink *sink) {
  const char *key = getenv("RESAS_API_KEY");
  if (!key || !*key) { fprintf(stderr, "[resas-municipality] gated (RESAS_API_KEY)\n"); return 0; }

  char authhdr[256];
  snprintf(authhdr, sizeof authhdr, "X-API-KEY: %s", key);
  const char *hdrs[] = { authhdr, NULL };

  int n = 0, fetched = 0;
  for (int pref = 1; pref <= 47; pref++) {
    char url[256];
    snprintf(url, sizeof url, "%s/cities?prefCode=%d", BASE, pref);
    cJSON *root = feed_get_json_h(ctx->http, url, hdrs, 12000);
    if (!root) continue;
    fetched++;
    cJSON *result = cJSON_GetObjectItem(root, "result");
    if (!result || !cJSON_IsArray(result)) { cJSON_Delete(root); continue; }

    cJSON *c;
    cJSON_ArrayForEach(c, result) {
      const char *cityCode = jo_vstr(cJSON_GetObjectItem(c, "cityCode"));
      const char *cityName = jo_vstr(cJSON_GetObjectItem(c, "cityName"));
      cJSON *bigFlag = cJSON_GetObjectItem(c, "bigCityFlag");
      const char *bigStr = jo_vstr(bigFlag);

      char rk[64], title[160], summary[160], body[256];
      snprintf(rk, sizeof rk, "%d-%s", pref, cityCode ? cityCode : "null");
      snprintf(title, sizeof title, "%s (pref %d)",
               cityName ? cityName : "municipality", pref);
      snprintf(summary, sizeof summary, "cityCode=%s bigCityFlag=%s",
               cityCode ? cityCode : "null", bigStr ? bigStr : "-");
      snprintf(body, sizeof body,
               "RESAS municipality: prefCode=%d, cityCode=%s, name=%s",
               pref, cityCode ? cityCode : "null",
               cityName ? cityName : "null");

      cJSON *p = cJSON_CreateObject();
      cJSON_AddNumberToObject(p, "prefCode", pref);
      if (cityCode) cJSON_AddStringToObject(p, "cityCode", cityCode);
      else cJSON_AddNullToObject(p, "cityCode");
      if (cityName) cJSON_AddStringToObject(p, "cityName", cityName);
      else cJSON_AddNullToObject(p, "cityName");
      if (bigFlag && !cJSON_IsNull(bigFlag))
        cJSON_AddItemToObject(p, "bigCityFlag", cJSON_Duplicate(bigFlag, 1));
      else cJSON_AddNullToObject(p, "bigCityFlag");
      char *pj = cJSON_PrintUnformatted(p);

      cJSON *tags = cJSON_CreateArray();
      cJSON_AddItemToArray(tags, cJSON_CreateString("statistics"));
      cJSON_AddItemToArray(tags, cJSON_CreateString("population"));
      cJSON_AddItemToArray(tags, cJSON_CreateString("municipality"));
      cJSON_AddItemToArray(tags, cJSON_CreateString("resas"));
      char *tj = cJSON_PrintUnformatted(tags);

      intel_item it = {0};
      it.remote_key      = rk;
      it.title           = title;
      it.summary         = summary;
      it.body            = body;
      it.link            = "https://opendata.resas-portal.go.jp/";
      it.lang            = "ja";
      it.record_type     = "resas-municipality";
      it.properties_json = pj;
      it.tags_json       = tj;
      if (sink->emit(sink, &it) >= 0) n++;

      free(pj); free(tj);
      cJSON_Delete(p); cJSON_Delete(tags);
    }
    cJSON_Delete(root);
  }
  fprintf(stderr, "[resas-municipality] emitted %d (%d/47 prefectures fetched)\n",
          n, fetched);
  /* STATUS code, not a row count: a prefecture with no new cities is an honest
   * empty. Only a total fetch failure is a real error. */
  return fetched > 0 ? 0 : -1;
}

static const source_def resas_municipality_def = {
  .id = "resas-municipality", .collector = "statistics",
  .name = "RESAS Municipal Comparison", .name_ja = "RESAS 自治体比較",
   .update_interval_sec = 86400, .run = run };
REGISTER_SOURCE(resas_municipality_def)
